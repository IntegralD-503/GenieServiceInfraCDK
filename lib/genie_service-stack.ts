import * as cdk from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Artifact } from 'aws-cdk-lib/aws-codepipeline';
import { Peer, Port, SecurityGroup, SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';
import { Repository } from 'aws-cdk-lib/aws-ecr';
import { DockerImageAsset } from 'aws-cdk-lib/aws-ecr-assets';
import { Cluster, ContainerImage, FargateService, FargateTaskDefinition, LogDrivers } from 'aws-cdk-lib/aws-ecs';
import { ApplicationLoadBalancer } from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { CfnParametersCode, Code } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
// import * as sqs from 'aws-cdk-lib/aws-sqs';

export class GenieServiceStack extends cdk.Stack {
  public readonly genieService: FargateService;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);


    // const serviceAsset = new DockerImageAsset(this, "GenieServiceAsset", {
    //   directory: "./app",
    //   file: "dockerfile"
    // })

    const vpc = new Vpc(this, "GenieServiceVPC", {
      cidr: "10.1.0.0/16",
      natGateways: 1,
      subnetConfiguration: [
        {  cidrMask: 24, subnetType: SubnetType.PUBLIC, name: "Public" },
        {  cidrMask: 24, subnetType: SubnetType.PRIVATE_WITH_EGRESS, name: "Private" }
        ],
      maxAzs: 3 // Default is all AZs in region
    });

    const repository = new Repository(this, "genie-service", {
      repositoryName: "genie-service"
    });


    const cluster = new Cluster(this, "GenieServiceCluster", {
      vpc: vpc
    });

    const executionRolePolicy =  new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['*'],
      actions: [
                "ecr:GetAuthorizationToken",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
            ]
    });
    
    const fargateTaskDefinition = new FargateTaskDefinition(this, 'GenieServiceTaskDefinition', {
      memoryLimitMiB: 4096,
      cpu: 512
      ,
    });
    fargateTaskDefinition.addToExecutionRolePolicy(executionRolePolicy);
    // fargateTaskDefinition.addToTaskRolePolicy(new iam.PolicyStatement({
    //   effect: iam.Effect.ALLOW,
    //   resources: [table.tableArn],
    //   actions: ['dynamodb:*']
    // }));
    
    const container = fargateTaskDefinition.addContainer("genie-service", {
      // Use an image from Amazon ECR
      image: ContainerImage.fromRegistry(repository.repositoryUri),
      logging: LogDrivers.awsLogs({streamPrefix: 'genie-service'}),
      environment: { 
        // 'DYNAMODB_MESSAGES_TABLE': table.tableName,
        'APP_ID' : 'genie-service'
      },
      
      // ... other options here ...
    });
    
    container.addPortMappings({
      containerPort: 3000
    });


    const sg_service = new SecurityGroup(this, 'MySGGenieService', { vpc: vpc });
    sg_service.addIngressRule(Peer.ipv4('0.0.0.0/0'), Port.tcp(3000));
    
    this.genieService = new FargateService(this, 'GenieFargateService', {
      cluster,
      taskDefinition: fargateTaskDefinition,
      desiredCount: 2,
      assignPublicIp: false,
      securityGroups: [sg_service]
    });
    
    // Setup AutoScaling policy
    const scaling = this.genieService.autoScaleTaskCount({ maxCapacity: 6, minCapacity: 2 });
    scaling.scaleOnCpuUtilization('CpuScaling', {
      targetUtilizationPercent: 50,
      scaleInCooldown: Duration.seconds(60),
      scaleOutCooldown: Duration.seconds(60)
    });


    const lb = new ApplicationLoadBalancer(this, 'GenieServiceALB', {
      vpc,
      internetFacing: true,
      
    });

    

    const listener = lb.addListener('Listener', {
      port: 80,
    });

    listener.addTargets('Target', {
      port: 80,
      targets: [this.genieService],
      healthCheck: { 
        path: '/health_check',
        // interval: Duration.seconds(120) 
      }
    });

    listener.connections.allowDefaultPortFromAnyIpv4('Open to the world');


  }
}
