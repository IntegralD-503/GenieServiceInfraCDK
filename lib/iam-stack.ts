import { Stack, StackProps } from "aws-cdk-lib";
import { Effect, ManagedPolicy, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { Construct } from "constructs";

export class IamStack extends Stack {

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        new ManagedPolicy(this, 'codeBuildECRTokenAccessPolicy', {
            statements: [
                new PolicyStatement({
                  effect: Effect.ALLOW,
                  resources: ['*'],
                  actions: [
                            "ecr:GetAuthorizationToken",
                        ]
                })
              ]
        })
        
    
        // const codeBuildECRTokenAccessRole = new Role(this, "GenieServiceBuildActionECRRole", {
        //     assumedBy: new ServicePrincipal("codebuild.amazonaws.com"),
        //     inlinePolicies: {
        //      CodeBuildECRTokenAccess: codeBuildECRTokenAccess
        //     }
        //   })
    
    
    }
}