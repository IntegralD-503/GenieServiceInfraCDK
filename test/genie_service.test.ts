import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as GenieService from '../lib/genie_service-stack';

// example test. To run these tests, uncomment this file along with the
// example resource in lib/genie_service-stack.ts
test('Genie Service Stack Created', () => {
  const app = new cdk.App();
    // WHEN
  const stack = new GenieService.GenieServiceStack(app, 'MyTestStack');
    // THEN
  const template = Template.fromStack(stack);

//   template.hasResourceProperties('AWS::ECR::Repository', {
//   });

  expect(template).toMatchSnapshot();
});
