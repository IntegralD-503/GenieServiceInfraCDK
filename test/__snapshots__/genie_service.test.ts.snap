// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Genie Service Stack Created 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GenieFargateService10EC5F4D": {
      "DependsOn": [
        "GenieServiceALBListenerBCA7A66E",
        "GenieServiceALBListenerTargetGroup61E00DC0",
      ],
      "Properties": {
        "Cluster": {
          "Ref": "GenieServiceCluster2E8C8F93",
        },
        "DeploymentConfiguration": {
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 50,
        },
        "DesiredCount": 2,
        "EnableECSManagedTags": false,
        "HealthCheckGracePeriodSeconds": 60,
        "LaunchType": "FARGATE",
        "LoadBalancers": [
          {
            "ContainerName": "genie-service",
            "ContainerPort": 3000,
            "TargetGroupArn": {
              "Ref": "GenieServiceALBListenerTargetGroup61E00DC0",
            },
          },
        ],
        "NetworkConfiguration": {
          "AwsvpcConfiguration": {
            "AssignPublicIp": "DISABLED",
            "SecurityGroups": [
              {
                "Fn::GetAtt": [
                  "MySGGenieServiceE43CDB86",
                  "GroupId",
                ],
              },
            ],
            "Subnets": [
              {
                "Ref": "GenieServiceVPCPrivateSubnet1Subnet490FE903",
              },
              {
                "Ref": "GenieServiceVPCPrivateSubnet2SubnetCB3682CE",
              },
            ],
          },
        },
        "TaskDefinition": {
          "Ref": "GenieServiceTaskDefinition8D07087D",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "GenieFargateServiceTaskCountTarget90792478": {
      "Properties": {
        "MaxCapacity": 6,
        "MinCapacity": 2,
        "ResourceId": {
          "Fn::Join": [
            "",
            [
              "service/",
              {
                "Ref": "GenieServiceCluster2E8C8F93",
              },
              "/",
              {
                "Fn::GetAtt": [
                  "GenieFargateService10EC5F4D",
                  "Name",
                ],
              },
            ],
          ],
        },
        "RoleARN": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::",
              {
                "Ref": "AWS::AccountId",
              },
              ":role/aws-service-role/ecs.application-autoscaling.amazonaws.com/AWSServiceRoleForApplicationAutoScaling_ECSService",
            ],
          ],
        },
        "ScalableDimension": "ecs:service:DesiredCount",
        "ServiceNamespace": "ecs",
      },
      "Type": "AWS::ApplicationAutoScaling::ScalableTarget",
    },
    "GenieFargateServiceTaskCountTargetCpuScaling38AB28AE": {
      "Properties": {
        "PolicyName": "MyTestStackGenieFargateServiceTaskCountTargetCpuScaling9717BA18",
        "PolicyType": "TargetTrackingScaling",
        "ScalingTargetId": {
          "Ref": "GenieFargateServiceTaskCountTarget90792478",
        },
        "TargetTrackingScalingPolicyConfiguration": {
          "PredefinedMetricSpecification": {
            "PredefinedMetricType": "ECSServiceAverageCPUUtilization",
          },
          "ScaleInCooldown": 60,
          "ScaleOutCooldown": 60,
          "TargetValue": 50,
        },
      },
      "Type": "AWS::ApplicationAutoScaling::ScalingPolicy",
    },
    "GenieServiceALB3318FABE": {
      "DependsOn": [
        "GenieServiceVPCPublicSubnet1DefaultRoute1E935F62",
        "GenieServiceVPCPublicSubnet1RouteTableAssociationBA1BE91D",
        "GenieServiceVPCPublicSubnet2DefaultRoute1D93F898",
        "GenieServiceVPCPublicSubnet2RouteTableAssociation4A17B59E",
      ],
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "GenieServiceALBSecurityGroup2AB7BC0C",
              "GroupId",
            ],
          },
        ],
        "Subnets": [
          {
            "Ref": "GenieServiceVPCPublicSubnet1Subnet9460CA9C",
          },
          {
            "Ref": "GenieServiceVPCPublicSubnet2Subnet2B18ACD2",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "GenieServiceALBListenerBCA7A66E": {
      "Properties": {
        "DefaultActions": [
          {
            "TargetGroupArn": {
              "Ref": "GenieServiceALBListenerTargetGroup61E00DC0",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "GenieServiceALB3318FABE",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "GenieServiceALBListenerTargetGroup61E00DC0": {
      "Properties": {
        "HealthCheckPath": "/health_check",
        "Port": 80,
        "Protocol": "HTTP",
        "TargetGroupAttributes": [
          {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "ip",
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "GenieServiceALBSecurityGroup2AB7BC0C": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB MyTestStackGenieServiceALBEAC5D04B",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "GenieServiceALBSecurityGrouptoMyTestStackMySGGenieService20B0E79B3000D6E4F1E5": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "MySGGenieServiceE43CDB86",
            "GroupId",
          ],
        },
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "GenieServiceALBSecurityGroup2AB7BC0C",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "GenieServiceCluster2E8C8F93": {
      "Type": "AWS::ECS::Cluster",
    },
    "GenieServiceTaskDefinition8D07087D": {
      "Properties": {
        "ContainerDefinitions": [
          {
            "Environment": [
              {
                "Name": "APP_ID",
                "Value": "genie-service",
              },
            ],
            "Essential": true,
            "Image": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Select": [
                      4,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "genieservice52C06942",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".dkr.ecr.",
                  {
                    "Fn::Select": [
                      3,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "genieservice52C06942",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".",
                  {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  {
                    "Ref": "genieservice52C06942",
                  },
                ],
              ],
            },
            "LogConfiguration": {
              "LogDriver": "awslogs",
              "Options": {
                "awslogs-group": {
                  "Ref": "GenieServiceTaskDefinitiongenieserviceLogGroup94883B48",
                },
                "awslogs-region": {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "genie-service",
              },
            },
            "Name": "genie-service",
            "PortMappings": [
              {
                "ContainerPort": 3000,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "Cpu": "512",
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "GenieServiceTaskDefinitionExecutionRole97A3AAC4",
            "Arn",
          ],
        },
        "Family": "MyTestStackGenieServiceTaskDefinition88D350CF",
        "Memory": "4096",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE",
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "GenieServiceTaskDefinitionTaskRoleF141B44F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "GenieServiceTaskDefinitionExecutionRole97A3AAC4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "GenieServiceTaskDefinitionExecutionRoleDefaultPolicyD062A287": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ecr:GetAuthorizationToken",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "GenieServiceTaskDefinitiongenieserviceLogGroup94883B48",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GenieServiceTaskDefinitionExecutionRoleDefaultPolicyD062A287",
        "Roles": [
          {
            "Ref": "GenieServiceTaskDefinitionExecutionRole97A3AAC4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GenieServiceTaskDefinitionTaskRoleF141B44F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "GenieServiceTaskDefinitiongenieserviceLogGroup94883B48": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "GenieServiceVPCD4E0FC4E": {
      "Properties": {
        "CidrBlock": "10.1.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "GenieServiceVPCIGW0CE02FDF": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "GenieServiceVPCPrivateSubnet1DefaultRoute6D062A53": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "GenieServiceVPCPublicSubnet1NATGateway28A7050E",
        },
        "RouteTableId": {
          "Ref": "GenieServiceVPCPrivateSubnet1RouteTableBB03880C",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "GenieServiceVPCPrivateSubnet1RouteTableAssociation6178E49B": {
      "Properties": {
        "RouteTableId": {
          "Ref": "GenieServiceVPCPrivateSubnet1RouteTableBB03880C",
        },
        "SubnetId": {
          "Ref": "GenieServiceVPCPrivateSubnet1Subnet490FE903",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "GenieServiceVPCPrivateSubnet1RouteTableBB03880C": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "GenieServiceVPCPrivateSubnet1Subnet490FE903": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.1.2.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "GenieServiceVPCPrivateSubnet2DefaultRoute9E400E9E": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "GenieServiceVPCPublicSubnet1NATGateway28A7050E",
        },
        "RouteTableId": {
          "Ref": "GenieServiceVPCPrivateSubnet2RouteTableF6ED56E8",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "GenieServiceVPCPrivateSubnet2RouteTableAssociation84FB22C4": {
      "Properties": {
        "RouteTableId": {
          "Ref": "GenieServiceVPCPrivateSubnet2RouteTableF6ED56E8",
        },
        "SubnetId": {
          "Ref": "GenieServiceVPCPrivateSubnet2SubnetCB3682CE",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "GenieServiceVPCPrivateSubnet2RouteTableF6ED56E8": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "GenieServiceVPCPrivateSubnet2SubnetCB3682CE": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.1.3.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "GenieServiceVPCPublicSubnet1DefaultRoute1E935F62": {
      "DependsOn": [
        "GenieServiceVPCVPCGW7A629517",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "GenieServiceVPCIGW0CE02FDF",
        },
        "RouteTableId": {
          "Ref": "GenieServiceVPCPublicSubnet1RouteTable6E2D8750",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "GenieServiceVPCPublicSubnet1EIPC0FF856D": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "GenieServiceVPCPublicSubnet1NATGateway28A7050E": {
      "DependsOn": [
        "GenieServiceVPCPublicSubnet1DefaultRoute1E935F62",
        "GenieServiceVPCPublicSubnet1RouteTableAssociationBA1BE91D",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "GenieServiceVPCPublicSubnet1EIPC0FF856D",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "GenieServiceVPCPublicSubnet1Subnet9460CA9C",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "GenieServiceVPCPublicSubnet1RouteTable6E2D8750": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "GenieServiceVPCPublicSubnet1RouteTableAssociationBA1BE91D": {
      "Properties": {
        "RouteTableId": {
          "Ref": "GenieServiceVPCPublicSubnet1RouteTable6E2D8750",
        },
        "SubnetId": {
          "Ref": "GenieServiceVPCPublicSubnet1Subnet9460CA9C",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "GenieServiceVPCPublicSubnet1Subnet9460CA9C": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.1.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "GenieServiceVPCPublicSubnet2DefaultRoute1D93F898": {
      "DependsOn": [
        "GenieServiceVPCVPCGW7A629517",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "GenieServiceVPCIGW0CE02FDF",
        },
        "RouteTableId": {
          "Ref": "GenieServiceVPCPublicSubnet2RouteTableD592EDED",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "GenieServiceVPCPublicSubnet2RouteTableAssociation4A17B59E": {
      "Properties": {
        "RouteTableId": {
          "Ref": "GenieServiceVPCPublicSubnet2RouteTableD592EDED",
        },
        "SubnetId": {
          "Ref": "GenieServiceVPCPublicSubnet2Subnet2B18ACD2",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "GenieServiceVPCPublicSubnet2RouteTableD592EDED": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "GenieServiceVPCPublicSubnet2Subnet2B18ACD2": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.1.1.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "MyTestStack/GenieServiceVPC/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "GenieServiceVPCVPCGW7A629517": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "GenieServiceVPCIGW0CE02FDF",
        },
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "MySGGenieServiceE43CDB86": {
      "Properties": {
        "GroupDescription": "MyTestStack/MySGGenieService",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:3000",
            "FromPort": 3000,
            "IpProtocol": "tcp",
            "ToPort": 3000,
          },
        ],
        "VpcId": {
          "Ref": "GenieServiceVPCD4E0FC4E",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "MySGGenieServicefromMyTestStackGenieServiceALBSecurityGroup6A2739F43000D5F12A88": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "MySGGenieServiceE43CDB86",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "GenieServiceALBSecurityGroup2AB7BC0C",
            "GroupId",
          ],
        },
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "genieservice52C06942": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RepositoryName": "genie-service",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
